src/hlsBinaries.ts

```diff
@@ -546,7 +546,11 @@ export async function upgradeGHCup(context: ExtensionContext, logger: Logger): P
   if (manageHLS === 'GHCup') {
     const upgrade = workspace.getConfiguration('haskell').get('upgradeGHCup') as boolean;
     if (upgrade) {
-      await callGHCup(context, logger, ['upgrade'], 'Upgrading ghcup', true);
+      try {
+        await callGHCup(context, logger, ['upgrade'], 'Upgrading ghcup', true);
+      } catch (e) {
+        window.showWarningMessage(`Failed to upgrade GHCup. Continuing anyway. You can disable upgrades via 'haskell.upgradeGHCup: false'`);
```

Comment:
Personally, I would even prefer if there was no warning popup. It would be great if the extension could detect that `ghcup` is installed using `brew` (or something else where `upgrade` is disabled) and then just not try to run `upgrade` / automatically default `haskell.upgradeGHCup` to `false`.

Reply:
That's unfortunate -- I have created [an issue upstream](https://gitlab.haskell.org/haskell/ghcup-hs/-/issues/355) for this.

Reply:
>  Therefore, haskell.upgradeGHCup being true by default is unexpected behaviour

Why would that be unexpected? Requiring users to manually upgrade ghcup to fix bugs in the extension will be problematic. Since the extension is now tied to ghcup and HLS bindists, bug can bubble up and we may need to force updates to ensure consistency.

Reply:
I think this is rather an argument to drop the `disable-upgrade` feature and not to anything special in brew. Since the default in ghcup is to **not** do in-place upgrades, this is safe.

Reply:
> I think this is a communication problem. The [prompt](https://github.com/haskell/vscode-haskell/blob/f83db667d389b02103dde063509920f929dafea9/src/hlsBinaries.ts#L204-L211) says "do you want me to use HLS in $PATH or do you want me to install it using ghcup?". The prompt doesn't ask or indicate that the extension will actually try to upgrade ghcup as well. So it is unexpected in terms of the messaging.

Well, there's only so much info we can put into a popup. The README can explain this more in-depth.

> In any case, unless I misunderstand what a ghcup upgrade entails, I would expect this extension to call ghcup --version and check whether it is a compatible version and warn me if it is incompatible, rather than assuming that only the latest version of ghcup is compatible, whatever that may be.

What do you mean with **compatible**? Are you suggesting we now keep a list of compatible ghcup versions around to support users who don't want to upgrade? GHCup is rolling release and we don't want to support older versions. The metadata makes this clear by [only ever listing one version](https://github.com/haskell/ghcup-metadata/blob/fbf06cc9eed096ba546530230f2facb0ac9815ba/ghcup-0.0.7.yaml#L2571). If a user doesn't upgrade on purpose, then that's on them.